src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java

```diff
@@ -60,25 +62,28 @@ public void rebuildSessionFromFrame(MenuSession menuSession) throws CommCareSess
                 EntityScreen entityScreen = (EntityScreen) screen;
                 entityScreen.init(menuSession.getSessionWrapper());
                 if (entityScreen.shouldBeSkipped()) {
-                    screen = menuSession.getNextScreen();
+                    screen = menuSession.getNextScreen(false);
                     continue;
                 }
                 SessionDatum neededDatum = entityScreen.getSession().getNeededDatum();
-                Set<String> entityIds = entityScreen.getReferenceMap().keySet();
                 for (StackFrameStep step: steps) {
                     if (step.getId().equals(neededDatum.getDataId())) {
-                        if (entityIds.contains(step.getValue())) {
-                            currentStep = step.getValue();
+                        for (TreeReference ref: entityScreen.getReferences()) {
+                            String id = entityScreen.getReturnValueFromSelection(ref,(EntityDatum) neededDatum, entityScreen.getEvalContext());
+                            if (id.equals(step.getValue())) {
+                                currentStep = step.getValue();
+                                break;
+                            }
                         }
                     }
```

Comment:
If I'm reading this correctly there should only be one step that matches the `neededDatum` right?
In that case I think this can be simplified (for readability) to:
```java
                Optional<StackFrameStep> matchingStep = steps.stream().
                        filter(step -> step.getId().equals(neededDatum.getDataId()))
                        .findFirst();
                if (matchingStep.isPresent()) {
                    String value = matchingStep.get().getValue();
                    for (TreeReference ref: entityScreen.getReferences()) {
                        String id = entityScreen.getReturnValueFromSelection(ref,(EntityDatum) neededDatum, entityScreen.getEvalContext());
                        if (id.equals(value)) {
                            currentStep = value;
                            break;
                        }
                    }
                }
```

Reply:
In theory, the app can be set up where two steps match `neededDatum` (`neededDatum.getDataId()` returns `case_id`), in cases where there are multiple case selections (parent/child selection in case lists). There are active QA bugs with that workflow right now (https://dimagi-dev.atlassian.net/browse/QA-2426?focusedCommentId=120648), but it is a possibility.

Reply:
in that case you could do this. It's a bit more complex but it prevents having to iterate over the list of references from the beginning each time and compute the return value. I don't know how much that gains us but if the reference list is long or the cost of computing the return value is not insignificant than it could be a reasonable gain.

```java
                List<String> neededIds = steps.stream().filter(step -> step.getId().equals(neededDatum.getDataId()))
                        .map(step -> step.getValue())
                        .collect(Collectors.toList());
                Iterator<TreeReference> references = entityScreen.getReferences().iterator();
                Set<String> seenIds = new HashSet<String>();
                for (String neededId : neededIds) {
                    if (seenIds.contains(neededId)){
                        currentStep = neededId;
                    } else {
                        while (references.hasNext()) {
                            TreeReference ref = references.next();
                            String refId = entityScreen.getReturnValueFromSelection(ref, (EntityDatum) neededDatum, entityScreen.getEvalContext());
                            seenIds.add(refId);
                            if (refId.equals(neededId)) {
                                currentStep = neededId;
                                break;
                            }
                        }
                    }
                }
```

Reply:
I think I was incorrect here. Whether or not there are multiple steps that match the needed datum doesn't really matter, since what we are checking is if the case id exists, so if there are multiple steps with the same id, the answer would be the same, since the datum is the same. A different datum, such as for parent/child selection, would imply that the second check would be in a different `EntityScreen`, so we would be reevaluating all of this, and would not be able to share results (the while loop wrapping all of this will have advanced).